internal/rukpak/preflights/crdupgradesafety/checks.go

```diff
@@ -70,3 +73,246 @@ func (c *ServedVersionValidator) Validate(old, new apiextensionsv1.CustomResourc
+	case diff.Old.Maximum != nil && diff.New.Maximum != nil && *diff.New.Maximum < *diff.Old.Maximum:
+		err = fmt.Errorf("maximum constraint decreased from %v to %v", *diff.Old.Maximum, *diff.New.Maximum)
+	}
+
+	return isHandled(diff, reset), err
+}
+
+func MaxItems(diff kappcus.FieldDiff) (bool, error) {
+	reset := func(diff kappcus.FieldDiff) kappcus.FieldDiff {
+		diff.Old.MaxItems = nil
+		diff.New.MaxItems = nil
+		return diff
+	}
+
+	var err error
+
+	switch {
+	case diff.Old.MaxItems == nil && diff.New.MaxItems != nil:
+		err = fmt.Errorf("maxItems constraint %v added when there were no restrictions previously", *diff.New.MaxItems)
+	case diff.Old.MaxItems != nil && diff.New.MaxItems != nil && *diff.New.MaxItems < *diff.Old.MaxItems:
+		err = fmt.Errorf("maxItems constraint decreased from %v to %v", *diff.Old.MaxItems, *diff.New.MaxItems)
+	}
+
+	return isHandled(diff, reset), err
+}
+
+func MaxLength(diff kappcus.FieldDiff) (bool, error) {
+	reset := func(diff kappcus.FieldDiff) kappcus.FieldDiff {
+		diff.Old.MaxLength = nil
+		diff.New.MaxLength = nil
+		return diff
+	}
+
+	var err error
+
+	switch {
+	case diff.Old.MaxLength == nil && diff.New.MaxLength != nil:
+		err = fmt.Errorf("maxLength constraint %v added when there were no restrictions previously", *diff.New.MaxLength)
+	case diff.Old.MaxLength != nil && diff.New.MaxLength != nil && *diff.New.MaxLength < *diff.Old.MaxLength:
+		err = fmt.Errorf("maxLength constraint decreased from %v to %v", *diff.Old.MaxLength, *diff.New.MaxLength)
+	}
+
+	return isHandled(diff, reset), err
+}
+
+func MaxProperties(diff kappcus.FieldDiff) (bool, error) {
+	reset := func(diff kappcus.FieldDiff) kappcus.FieldDiff {
+		diff.Old.MaxProperties = nil
+		diff.New.MaxProperties = nil
+		return diff
+	}
+
+	var err error
+
+	switch {
+	case diff.Old.MaxProperties == nil && diff.New.MaxProperties != nil:
+		err = fmt.Errorf("maxProperties constraint %v added when there were no restrictions previously", *diff.New.MaxProperties)
+	case diff.Old.MaxProperties != nil && diff.New.MaxProperties != nil && *diff.New.MaxProperties < *diff.Old.MaxProperties:
+		err = fmt.Errorf("maxProperties constraint decreased from %v to %v", *diff.Old.MaxProperties, *diff.New.MaxProperties)
+	}
+
+	return isHandled(diff, reset), err
+}
+
+func Minimum(diff kappcus.FieldDiff) (bool, error) {
+	reset := func(diff kappcus.FieldDiff) kappcus.FieldDiff {
+		diff.Old.Minimum = nil
+		diff.New.Minimum = nil
+		return diff
+	}
+
+	var err error
+
+	switch {
+	case diff.Old.Minimum == nil && diff.New.Minimum != nil:
+		err = fmt.Errorf("minimum constraint %v added when there were no restrictions previously", *diff.New.Minimum)
+	case diff.Old.Minimum != nil && diff.New.Minimum != nil && *diff.New.Minimum > *diff.Old.Minimum:
+		err = fmt.Errorf("minimum constraint increased from %v to %v", *diff.Old.Minimum, *diff.New.Minimum)
+	}
```

Comment:
This switch statement is for the most part repeated across all the `Min*` functions. I'm planning to make an update to have a function that is re-used across all of them, but didn't have time left and wanted to get this PR up before I logged off for the day. I figure it's not the end of the world if this goes in as-is since I could file a follow-up. If it hasn't merged by the time I get to circle back around to it I'll push the update :)

Reply:
Done in https://github.com/operator-framework/operator-controller/pull/1269/commits/787b14815e628f68fae683e2125f6b4289dd380e